### PR TITLE
Adds an autockick subsystem

### PR DESCRIPTION
**Pull request by Gandalf2k15**
## About The Pull Request

Adds a new subsystem for kicking people automatically depending on the threshold. This will only ever apply to someone that is AFK.
**Critical:**
Kicks all AFK people, regardless of state.
**Moderate:**
Kicks all AFK people that are observing/in the lobby.
**Normal:**
Kicks all AFK people in the lobby.

Adds two new admin commands:
Kick ALL AFK clients: Gives you the choice to kick everyone or just the lobby/observing.
Move ghosts to lobby: Allows you to move all ghosts to lobby and adds them to the cleanup list of the AFK SS, they must either join or leave.

## Changelog

:cl:
add: Added two new verbs for admins, kick all and move ghosts to lobby.
add: Added a new AFK subsystem that will automatically kick anyone that is AFK depending on the severity of the situation.
/:cl:

### CODE REVIEW

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	var/kick_type_choice = tgui_alert("Do you want to kick all AFK clients, or only those observing or in the lobby?", "All of them", "Only those in the lobby/observing", "Cancel")
+
+	if(!kick_type_choice || kick_type_choice == "Cancel")
+		return
+
+	var/list/kicked_client_names = list()
+
+	if(kick_type_choice == "All of them")
+		for(var/client/iterating_client as anything in GLOB.clients)
+			if(iterating_client.is_afk())
```
Comment (jjpark-kb) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]:
```suggestion
		for(var/client/iterating_client as anything in GLOB.clients)
			if(iterating_client?.is_afk())
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+				to_chat_immediate(iterating_client, "You have been kicked for being AFK.")
+				kicked_client_names.Add("[iterating_client.key]")
+				qdel(iterating_client)
+		message_admins("[key_name(usr)] has kicked ALL AFK clients. Kicked [kicked_client_names.len] players.")
+	else
+		for(var/client/iterating_client as anything in GLOB.clients)
+			if(iterating_client.is_afk() && isnewplayer(iterating_client.mob) || isobserver(iterating_client.mob))
```
Comment (jjpark-kb) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]:
```suggestion
		for(var/client/iterating_client as anything in GLOB.clients)
			if(iterating_client?.is_afk() && isnewplayer(iterating_client.mob) || isobserver(iterating_client.mob))
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm (hunk 1 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/antilag/code/admin_procs.dm]: or maybe just an if(!iterating_client)

File: modular_skyrat/modules/antilag/code/admin_procs.dm (hunk 2 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/antilag/code/admin_procs.dm]: or maybe just an if(!iterating_client)

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+	var/choice = tgui_alert("Are you sure you want to move all ghosts to the lobby?", "Yes", "No")
+
+	if(!choice || choice == "No")
+		return
+
+	var/list/ghost_client_names = list()
+
+	for(var/client/iterating_client as anything in GLOB.clients)
+		if(isobserver(iterating_client.mob))
```
Comment (jjpark-kb) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]: maybe an if(!iterating_client)

File: modular_skyrat/modules/antilag/code/admin_procs.dm (hunk 1 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/antilag/code/admin_procs.dm]: that global list should never have a null in it

Comment (jjpark-kb) [reply modular_skyrat/modules/antilag/code/admin_procs.dm]: Alright, just making sure.

Comment (Gandalf2k15) [reply modular_skyrat/modules/antilag/code/admin_procs.dm]: Probably good to have a check tho

**Comment by jjpark-kb** [comment]
how dare you try to remove community members from the server that are AFK (not contributing to the current round) in order to resolve any issues regarding performance quality, etc.!

**Comment by nikothedude** [comment]
my beloved 👍

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+	if(kick_type_choice == "All of them")
+		for(var/client/iterating_client as anything in GLOB.clients)
+			if(!iterating_client)
+				continue
+			if(iterating_client.is_afk())
+				to_chat_immediate(iterating_client, "You have been kicked for being AFK.")
+				kicked_client_names.Add("[iterating_client.key]")
+				qdel(iterating_client)
+		message_admins("[key_name(usr)] has kicked ALL AFK clients. Kicked [kicked_client_names.len] players.")
```
Comment (jjpark-kb) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]:
```suggestion
		message_admins("[key_name(usr)] has kicked ALL AFK clients. Kicked [length(kicked_client_names)] players.")
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	else
+		for(var/client/iterating_client as anything in GLOB.clients)
+			if(!iterating_client)
+				continue
+			if(iterating_client.is_afk() && isnewplayer(iterating_client.mob) || isobserver(iterating_client.mob))
+				to_chat_immediate(iterating_client, "You have been kicked for being AFK.")
+				kicked_client_names.Add("[iterating_client.key]")
+				qdel(iterating_client)
+		message_admins("[key_name(usr)] has kicked all AFK clients in the lobby/observing. Kicked [kicked_client_names.len] players.")
```
Comment (jjpark-kb) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]:
```suggestion
		message_admins("[key_name(usr)] has kicked all AFK clients in the lobby/observing. Kicked [length(kicked_client_names)] players.")
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+	to_chat(usr, span_admin("Total kicked clients: [kicked_client_names.len] | Kicked clients: " + kicked_client_names.Join(", ")))
```
Comment (jjpark-kb) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]:
```suggestion
	to_chat(usr, span_admin("Total kicked clients: [length(kicked_client_names)] | Kicked clients: " + kicked_client_names.Join(", ")))
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+			continue
+		if(isobserver(iterating_client.mob))
+			to_chat(iterating_client, span_danger("You have been moved to the lobby, either join a game or disconnect. You will shortly be kicked."))
+			SSautockick.clients_to_check_lobby.Add(iterating_client)
+			var/mob/dead/new_player/new_player = new()
+			new_player.ckey = iterating_client.ckey
+			ghost_client_names.Add("[iterating_client.key]")
+			qdel(iterating_client.mob)
+
+	message_admins("[key_name(usr)] has moved [ghost_client_names.len] ghosts to the lobby.")
```
Comment (jjpark-kb) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]:
```suggestion
	message_admins("[key_name(usr)] has moved [length(ghost_client_names)] ghosts to the lobby.")
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+	to_chat(usr, span_admin("Total moved observers: [ghost_client_names.len] | Moved observers: " + ghost_client_names.Join(", ")))
```
Comment (jjpark-kb) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]:
```suggestion
	to_chat(usr, span_admin("Total moved observers: [length(ghost_client_names)] | Moved observers: " + ghost_client_names.Join(", ")))
```

File: modular_skyrat/modules/antilag/code/kick_subsystem.dm [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+	name = "Autokick"
+	init_order = INIT_ORDER_TIMER
+	runlevels = RUNLEVEL_SETUP | RUNLEVEL_GAME
+	wait = DEFAULT_AUTOKICK_SS_WAIT
+
+	/// A list of clients we have been handed to check for lobby autokicking.
+	var/list/clients_to_check_lobby = list()
+
+/datum/controller/subsystem/autockick/stat_entry(msg)
+	msg += "|CHK:[clients_to_check_lobby.len]|TIME:[next_fire - world.time]"
```
Comment (jjpark-kb) [inline modular_skyrat/modules/antilag/code/kick_subsystem.dm]:
```suggestion
	msg += "|CHK:[length(clients_to_check_lobby)]|TIME:[next_fire - world.time]"
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+/client/proc/move_ghost_to_lobby()
+	set category = "Admin"
+	set name = "Move ghosts to lobby"
+	set desc = "Moves any ghosts to the lobby."
+
+	if(GLOB.clients.len < 150)
+		to_chat(usr, span_danger("There are not enough players to move ghosts to the lobby."))
+		return
+
+	var/choice = tgui_alert("Are you sure you want to move all ghosts to the lobby?", "Yes", "No")
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]:
```suggestion
	if(!(choice == "Yes"))
```
Slightly cleaner

File: modular_skyrat/modules/antilag/code/kick_subsystem.dm [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+				if(AUTOKICK_SEVERITY_NORMAL)
+					if(isnewplayer(iterating_client.mob)) // Kicking anyone in the lobby.
+						kicked_lobby += "[iterating_client.key]"
+						kick_player(iterating_client)
+						continue
+		if(MC_TICK_CHECK)
+			message_admins("AUTOKICK: MC check failed, aborting.")
+			return
+	var/total = kicked_lobby.len + kicked_ghosts.len + kicked_clients.len
+	message_admins("AUTOKICK: Kicked [total] total clients | Lobby: [kicked_lobby.len] | Ghosts: [kicked_ghosts.len] | Clients: [kicked_clients.len]")
```
Comment (jjpark-kb) [inline modular_skyrat/modules/antilag/code/kick_subsystem.dm]:
```suggestion
	var/total = length(kicked_lobby)+ length(kicked_ghosts)+ length(kicked_clients)
	message_admins("AUTOKICK: Kicked [total] total clients | Lobby: [length(kicked_lobby)] | Ghosts: [length(kicked_ghosts)] | Clients: [length(kicked_clients)]")
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+	var/kick_type_choice = tgui_alert("Do you want to kick all AFK clients, or only those observing or in the lobby?", "All of them", "Only those in the lobby/observing", "Cancel")
+
+	if(!kick_type_choice || kick_type_choice == "Cancel")
+		return
+
+	var/list/kicked_client_names = list()
+
+	if(kick_type_choice == "All of them")
+		for(var/client/iterating_client as anything in GLOB.clients)
+			if(!iterating_client)
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/admin_procs.dm:15]: What is the point in this

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+				continue
+			if(is_admin(iterating_client))
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/admin_procs.dm:17]:
```suggestion
			if(iterating_client?.is_afk())
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+				continue
+			if(iterating_client.is_afk())
+				to_chat_immediate(iterating_client, "You have been kicked for being AFK.")
+				kicked_client_names.Add("[iterating_client.key]")
+				qdel(iterating_client)
+		message_admins("[key_name(usr)] has kicked ALL AFK clients. Kicked [length(kicked_client_names)] players.")
+	else
+		for(var/client/iterating_client as anything in GLOB.clients)
+			if(!iterating_client)
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/admin_procs.dm:27]:
```suggestion
			if(iterating_client?.is_afk() && isnewplayer(iterating_client.mob) || isobserver(iterating_client.mob))
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+	var/choice = tgui_alert("Are you sure you want to move all ghosts to the lobby?", "Yes", "No")
+
+	if(!choice || choice == "No")
+		return
+
+	var/list/ghost_client_names = list()
+
+	for(var/client/iterating_client as anything in GLOB.clients)
+		if(!iterating_client)
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/admin_procs.dm:51]:
```suggestion
		if(isobserver(iterating_client?.mob))
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+			continue
+		if(is_admin(iterating_client))
+			continue
+		if(isobserver(iterating_client.mob))
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]:
```suggestion
			new_player.ckey = iterating_client?.ckey
```

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+			to_chat(iterating_client, span_danger("You have been moved to the lobby, either join a game or disconnect. You will shortly be kicked."))
+			SSautockick.clients_to_check_lobby.Add(iterating_client)
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/admin_procs.dm]:
```suggestion
			qdel(iterating_client?.mob)
```

File: modular_skyrat/modules/antilag/code/kick_subsystem.dm [hunk 18]
```diff
@@ -0,0 +1,64 @@
+SUBSYSTEM_DEF(autockick)
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/kick_subsystem.dm]: Am I missing something? Why is it `autockick` and not just `autokick`? Seems a bit confusing

File: modular_skyrat/modules/antilag/code/kick_subsystem.dm [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+	message_admins("AUTOKICK: Running AFK check... Population: [TGS_CLIENT_COUNT] | PRIORITY: [severity]")
+	for(var/client/iterating_client as anything in GLOB.clients) //Copied code is there for performance, why run two for loops when we can run one?
+		if(!iterating_client)
+			continue
+		if(is_admin(iterating_client))
+			continue
+		if((iterating_client in clients_to_check_lobby) && isnewplayer(iterating_client.mob))
+			clients_to_check_lobby -= iterating_client
+			to_chat_immediate(iterating_client, "As you have not joined the game, you have been kicked.")
+			qdel(iterating_client)
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/kick_subsystem.dm:39]:
```suggestion
		if(iterating_client?.is_afk())
```

File: modular_skyrat/modules/antilag/code/kick_subsystem.dm [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+			continue
+		if(iterating_client.is_afk())
+			switch(severity)
+				if(AUTOKICK_SEVERITY_CRITICAL)
+					kicked_clients += "[iterating_client.key]"
+					kick_player(iterating_client)
+					continue
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/kick_subsystem.dm:46]:
```suggestion
					if(isobserver(iterating_client?.mob))
```

File: modular_skyrat/modules/antilag/code/kick_subsystem.dm [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+				if(AUTOKICK_SEVERITY_MODERATE)
+					if(isobserver(iterating_client.mob))
+						kicked_ghosts += "[iterating_client.key]"
+						kick_player(iterating_client)
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/kick_subsystem.dm:50]:
```suggestion
					if(isnewplayer(iterating_client?.mob)) // Kicking anyone in the lobby.
```

File: modular_skyrat/modules/antilag/code/kick_subsystem.dm [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+						continue
+					if(isnewplayer(iterating_client.mob)) // Kicking anyone in the lobby.
+						kicked_lobby += "[iterating_client.key]"
+						kick_player(iterating_client)
+						continue
```
Comment (tf-4) [inline modular_skyrat/modules/antilag/code/kick_subsystem.dm:55]:
```suggestion
					if(isnewplayer(iterating_client?.mob)) // Kicking anyone in the lobby.
```

**Review by tf-4 — COMMENTED** [review]
Mostly good, just funny client null checks

File: modular_skyrat/modules/antilag/code/admin_procs.dm (hunk 12 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/antilag/code/admin_procs.dm:15]: Clients can be deleted at any moment and that list is liable to have a null in it.

File: modular_skyrat/modules/antilag/code/kick_subsystem.dm (hunk 18 above)
Comment (Gandalf2k15) [reply modular_skyrat/modules/antilag/code/kick_subsystem.dm]: Hmm, true

File: config/skyrat/skyrat_config.txt [hunk 23]
```diff
@@ -72,3 +72,6 @@ PC_MOB_TEXT As a player controlled mob you are expected to play the role to the
 ## Disables any ERP preferences for codebases that don't want it.
 #DISABLE_ERP_PREFERENCES
 #ERP_EMOTES_TO_DISABLE cum
+
+## Disables the automatic kicking system
+#DISABLE_AFK_AUTOKICK
```
Comment (GoldenAlpharex) [inline config/skyrat/skyrat_config.txt]: This should probably be "ENABLE_AFK_AUTOKICK" or something, negative variable names are usually not the best way to go around doing things.

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+		for(var/client/iterating_client as anything in GLOB.clients)
+			if(!iterating_client)
+				continue
+			if(is_admin(iterating_client))
+				continue
+			if(iterating_client.is_afk())
+				to_chat_immediate(iterating_client, "You have been kicked for being AFK.")
+				kicked_client_names.Add("[iterating_client.key]")
+				qdel(iterating_client)
+		message_admins("[key_name(usr)] has kicked ALL AFK clients. Kicked [length(kicked_client_names)] players.")
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/antilag/code/admin_procs.dm:23]: This should probably also be logged.

File: modular_skyrat/modules/antilag/code/admin_procs.dm [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+	else
+		for(var/client/iterating_client as anything in GLOB.clients)
+			if(!iterating_client)
+				continue
+			if(iterating_client.is_afk() && isnewplayer(iterating_client.mob) || isobserver(iterating_client.mob))
+				to_chat_immediate(iterating_client, "You have been kicked for being AFK.")
+				kicked_client_names.Add("[iterating_client.key]")
+				qdel(iterating_client)
+		message_admins("[key_name(usr)] has kicked all AFK clients in the lobby/observing. Kicked [length(kicked_client_names)] players.")
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/antilag/code/admin_procs.dm:33]: Ditto.

File: modular_skyrat/modules/antilag/code/admin_procs.dm (hunk 10 above)
Comment (GoldenAlpharex) [reply modular_skyrat/modules/antilag/code/admin_procs.dm]: I think you made your suggestion on the wrong line.

File: modular_skyrat/modules/antilag/code/admin_procs.dm (hunk 17 above)
Comment (GoldenAlpharex) [reply modular_skyrat/modules/antilag/code/admin_procs.dm]: why is this spread over like 6 comments lmao

File: modular_skyrat/modules/antilag/code/kick_subsystem.dm [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+	/// A list of clients we have been handed to check for lobby autokicking.
+	var/list/clients_to_check_lobby = list()
+
+/datum/controller/subsystem/autokick/stat_entry(msg)
+	msg += "|CHK:[length(clients_to_check_lobby)]|TIME:[next_fire - world.time]"
+	return ..()
+
+/datum/controller/subsystem/autokick/Initialize(start_timeofday)
+	if(CONFIG_GET(flag/ssdecay_disabled))
+		flags = SS_NO_FIRE
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/antilag/code/kick_subsystem.dm]: Bro you're not even checking the right config ;)

File: modular_skyrat/modules/antilag/code/kick_subsystem.dm [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+				if(AUTOKICK_SEVERITY_NORMAL)
+					if(isnewplayer(iterating_client.mob)) // Kicking anyone in the lobby.
+						kicked_lobby += "[iterating_client.key]"
+						kick_player(iterating_client)
+						continue
+		if(MC_TICK_CHECK)
+			message_admins("AUTOKICK: MC check failed, aborting.")
+			return
+	var/total = length(kicked_lobby)+ length(kicked_ghosts)+ length(kicked_clients)
+	message_admins("AUTOKICK: Kicked [total] total clients | Lobby: [length(kicked_lobby)] | Ghosts: [length(kicked_ghosts)] | Clients: [length(kicked_clients)]")
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/antilag/code/kick_subsystem.dm]: Add logs too (unless there's some in message_admins())

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
I'm tired, here's a tired review.

File: config/skyrat/skyrat_config.txt (hunk 23 above)
Comment (Gandalf2k15) [reply config/skyrat/skyrat_config.txt]: Hmm, we have a habit of leaving config changes out :)

**Comment by ORCACommander** [comment]
i coulda sworn i i saw auto kick controls in the tg configs and there are game panel buttons for kicking afk'er in lobby at least

**Comment by Gandalf2k15** [comment]
> i coulda sworn i i saw auto kick controls in the tg configs and there are game panel buttons for kicking afk'er in lobby at least

trhwer is but we want more fine control and broader spoecturm

**Comment by ORCACommander** [comment]
makes sense

**Comment by Gandalf2k15** [comment]
Plan A has failed lads, its time for plan B

**Comment by ghost** [comment]
wtf is plan b????

**Comment by Zonespace27** [comment]
> wtf is plan b????

A pop cap, #9979

**Comment by Gandalf2k15** [comment]
We'll be using this in combination with the crosslink

**Pull request closed without merging**